src/systems/gameplay_state_system.rs

```diff
@@ -26,17 +28,25 @@ impl<'a> System<'a> for GameplayStateSystem {
     fn run(&mut self, data: Self::SystemData) {
         let (mut game_state,
             mut sound_lib,
+            level_data,
             boxes,
             spots,
             renderables) = data;
 
         if game_state.gameplay_state != GameplayState::Won {
+            let none_string = "None";
             let spot_positions = (&spots, &renderables).join()
-                .map(|(_, renderable)| (renderable.position.x, renderable.position.y, &renderable.resource_template_data["spot_color"]))
+                .map(|(_, renderable)| (renderable.position.x, renderable.position.y, match level_data.box_spot_identical_mode {
```

Comment:
The logic is not good. Try other way to process with 2 different modes.
Consider hash_map over hash_set for better performance.